platform/ios/src/MGLMapView.mm

```diff
@@ -4150,6 +4173,111 @@ - (void)applyIconIdentifier:(NSString *)iconIdentifier toAnnotationsWithImageReu
+    
+    [self deselectShapeAnnotation:_selectedShapeAnnotation];
+    
+    _selectedShapeAnnotation = shapeAnnotation;
+    
+    if ([self.delegate respondsToSelector:@selector(mapView:didSelectShapeAnnotation:)])
+    {
+        [self.delegate mapView:self didSelectShapeAnnotation:shapeAnnotation];
+    }
+}
+
+- (void)deselectShapeAnnotation:(MGLShape *)shapeAnnotation
+{
+    if (!shapeAnnotation) return;
+    
+    if (_selectedShapeAnnotation == shapeAnnotation)
+    {
+        if ([self.delegate respondsToSelector:@selector(mapView:didDeselectShapeAnnotation:)])
+        {
+            [self.delegate mapView:self didDeselectShapeAnnotation:shapeAnnotation];
+        }
+        _selectedShapeAnnotation = nil;
+    }
+   
+}
+
+- (MGLShape*)shapeAnnotationForGestureRecognizer:(UITapGestureRecognizer*)singleTap
+{
+    CGPoint tapPoint = [singleTap locationInView:self];
+//    CGRect queryRect = CGRectInset({ tapPoint, CGSizeZero },
+//                                   -_unionedAnnotationRepresentationSize.width,
+//                                   -_unionedAnnotationRepresentationSize.height);
+//    queryRect = CGRectInset(queryRect, -MGLAnnotationImagePaddingForHitTest,
+//                            -MGLAnnotationImagePaddingForHitTest);
+//    
+//    NSMutableSet *layerIDs = [NSMutableSet set];
+//    
+//    for (const auto &annotation : _shapeAnnotationLayerIDs) {
+//        const std::string layerID = annotation.second;
+//        [layerIDs addObject:[NSString stringWithUTF8String:layerID.c_str()]];
+//    }
+//    
+//    NSArray *nearbyFeatures = [self visibleFeaturesInRect:queryRect inStyleLayersWithIdentifiers:layerIDs predicate:nil];
+//    id<MGLFeature> feature = nearbyFeatures.firstObject;
+//    
+    MGLAnnotationTag hitAnnotationTag = [self shapeAnnotationTagAtPoint:tapPoint];
+    
+    if (hitAnnotationTag != MGLAnnotationTagNotFound) {
+        id <MGLAnnotation> annotation = [self annotationWithTag:hitAnnotationTag];
+        NSAssert(annotation, @"Cannot select nonexistent annotation with tag %u", hitAnnotationTag);
+        if ([annotation isKindOfClass:[MGLShape class]]) {
+            return (MGLShape *)annotation;
+        }
+    }
+    
+    return nil;
+}
+
+- (MGLAnnotationTag)shapeAnnotationTagAtPoint:(CGPoint)point
+{
+    CGRect queryRect = CGRectInset({ point, CGSizeZero },
+                                   -_unionedAnnotationRepresentationSize.width,
+                                   -_unionedAnnotationRepresentationSize.height);
+    queryRect = CGRectInset(queryRect, -MGLAnnotationImagePaddingForHitTest,
+                            -MGLAnnotationImagePaddingForHitTest);
+    
+    std::vector<MGLAnnotationTag> nearbyAnnotations = [self shapeAnnotationTagsInRect:queryRect];
+    
+    MGLAnnotationTag hitAnnotationTag = MGLAnnotationTagNotFound;
+    
+    // Choose the first nearby annotation.
+    if (nearbyAnnotations.size())
+    {
+        hitAnnotationTag = nearbyAnnotations.front();
```

Comment:
The older -[MGLMapView annotationTagAtPoint:persistingResults:] method has logic to cycle through points that re overlapping / nearby each other. Do we need to worry about that for shapes? What is the expected behavior if a user touches in an area where several polygons overlap or several polylines intersect? cc @1ec5

Reply:
It’s currently possible to disable touch interaction for a given point annotation using either `MGLAnnotationImage.enabled`. That’s a weird API, but do we need something like it for shape annotations?